sdk/storage/azure-storage-blob/azure/storage/blob/_download.py

```diff
@@ -625,23 +654,50 @@ def read(self, size: Optional[int] = -1) -> T:
                 progress_hook=self._progress_hook,
                 **self._request_options
             )
-
-            if parallel and remaining_size > self._config.max_chunk_get_size:
-                import concurrent.futures
-                with concurrent.futures.ThreadPoolExecutor(self._max_concurrency) as executor:
-                    list(executor.map(
+            self._first_chunk = False
+
+            # When reading all data, have the downloader read everything into the stream.
+            # Else, read one chunk at a time (using the downloader as an iterator) until
+            # the requested size is reached.
+            chunks_iter = downloader.get_chunk_offsets()
+            if readall and not self._text_mode:
+                # Only do parallel if there is more than one chunk left to download
+                if parallel and (self.size - self._download_offset) > self._config.max_chunk_get_size:
+                    import concurrent.futures
+                    with concurrent.futures.ThreadPoolExecutor(self._max_concurrency) as executor:
+                        list(executor.map(
                             with_current_context(downloader.process_chunk),
                             downloader.get_chunk_offsets()
                         ))
-            else:
-                for chunk in downloader.get_chunk_offsets():
-                    downloader.process_chunk(chunk)
+                else:
+                    for chunk in chunks_iter:
+                        downloader.process_chunk(chunk)
 
-            self._offset += remaining_size
+                self._download_offset = self.size
+                self._offset = self.size
+
+            else:
+                while (chunk := next(chunks_iter, None)) is not None and remaining > 0:
+                    chunk_data = downloader.yield_chunk(chunk)
+                    self._download_offset += len(chunk_data)
+                    final = self._download_offset >= self.size
+                    self._current_content = self._decoder.decode(chunk_data, final) if self._text_mode else chunk_data
+                    if remaining < len(self._current_content):
```

Comment:
Is there a difference in how we track progress when using bytes vs chars? They wouldn't be quite the same right?

Reply:
Will ping offline about this one actually.